### PR TITLE
compiler: Introduce linearization pass

### PR DESCRIPTION
**Pull request by FabioLuporini**
I also took the opportunity of rewriting bits of the IET layer for a simpler future

### CODE REVIEW

File: devito/ir/iet/nodes.py [hunk 1]
```diff
@@ -611,19 +607,8 @@ def free_symbols(self):
 
     @property
     def defines(self):
-        """All Symbols defined in the Iteration header."""
         return self.dimensions
 
-    @property
```
Comment (georgebisbas) [inline devito/ir/iet/nodes.py:617]: was that dead code? Or moved else?

File: examples/compiler/03_iet-A.ipynb [hunk 2]
```diff
@@ -201,24 +201,7 @@
...
-      "        <TimedList (1, 1, 1)>\n",
-      "          <[affine,parallel,parallel=,skewable,tilable] Iteration x::x::(x_m, x_M, 1)>\n",
-      "            <[affine,parallel,parallel=,skewable,tilable] Iteration y::y::(y_m, y_M, 1)>\n",
-      "              <ExpressionBundle (1)>\n",
-      "\n",
-      "                <Expression u[t1, x + 1, y + 1] = u[t0, x + 1, y + 1] + 1>\n",
-      "\n",
-      "\n",
-      "\n"
+      "  <CallableBody>\n"
```
Comment (georgebisbas) [inline examples/compiler/03_iet-A.ipynb:204]: Need to see IET, right? Now useful info is hidden

File: devito/ir/iet/nodes.py (hunk 1 above)
Comment (FabioLuporini) [reply devito/ir/iet/nodes.py:617]: dead code

File: devito/core/arm.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 from devito.core.cpu import Cpu64AdvOperator
 from devito.passes.iet import (CTarget, OmpTarget, mpiize, optimize_halospots,
-                               hoist_prodders, relax_incr_dimensions)
+                               hoist_prodders, linearize, relax_incr_dimensions)
```
Comment (mloubout) [inline devito/core/arm.py:3]: Should this be a bit more explicit, like `linearize_index`

File: devito/core/operator.py [hunk 4]
```diff
@@ -195,6 +195,10 @@ def _specialize_iet(cls, graph, **kwargs):
         # Symbol definitions
         cls._Target.DataManager(sregistry, options).process(graph)
 
+        # Linearize n-dimensional Indexeds
+        if passes_mapper['linearize'] not in applied and options['linearize']:
+            passes_mapper['linearize'](graph)
```
Comment (mloubout) [inline devito/core/operator.py:200]: should it be added to `applied` then?

File: devito/ir/iet/nodes.py [hunk 5]
```diff
@@ -124,20 +124,20 @@ def args_frozen(self):
     def __str__(self):
         return str(self.ccode)
 
-    @abc.abstractproperty
+    @property
     def functions(self):
         """All AbstractFunction objects used by this node."""
-        raise NotImplementedError()
+        return ()
```
Comment (mloubout) [inline devito/ir/iet/nodes.py:130]: why the change?

File: devito/ir/iet/nodes.py [hunk 6]
```diff
@@ -689,13 +677,54 @@ def __repr__(self):
...
+        self.init = as_tuple(init)
+        self.unpacks = as_tuple(unpacks)
+        self.allocs = as_tuple(allocs)
+        self.casts = as_tuple(casts)
+        self.maps = as_tuple(maps)
+        self.unmaps = as_tuple(unmaps)
+        self.frees = as_tuple(frees)
+
+    def __repr__(self):
+        return "<CallableBody>"
```
Comment (mloubout) [inline devito/ir/iet/nodes.py:727]: No extra details like `self.body`?

File: devito/ir/iet/nodes.py [hunk 7]
```diff
@@ -1061,28 +1084,54 @@ def periodic(self):
...
-        self._free_symbols = as_tuple(free_symbols)
+    def __init__(self, pragmas):
+        super().__init__()
+        self.pragmas = as_tuple(pragmas)
 
-    @property
-    def pragmas(self):
-        return self.header
+    def __repr__(self):
+        return '<Pragmas>'
```
Comment (mloubout) [inline devito/ir/iet/nodes.py:1098]: `+self.pragmas`?

Comment (mloubout) [reply devito/ir/iet/nodes.py:1098]: May need `__str__ = __repr__`

File: devito/passes/iet/linearization.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+from devito.types import Symbol, FIndexed, Indexed, Wildcard
+from devito.types.basic import IndexedData
+
+
+__all__ = ['linearize']
+
+
+def linearize(graph, **kwargs):
+    """
+    Turn n-dimensional Indexeds into 1-dimensional Indexed with suitable index
```
Comment (mloubout) [inline devito/passes/iet/linearization.py:22]: May wanna specify is row major?

File: devito/symbolics/extended_sympy.py [hunk 9]
```diff
@@ -410,14 +422,14 @@ class Literal(sympy.Symbol):
...
-    np.int: INT,
+    int: INT,
     np.int32: INT,
     np.int64: INT,
     np.float32: FLOAT,
     float: DOUBLE,
     np.float64: DOUBLE,
 
-    (np.int, '*'): INTP,
+    (int, '*'): INTP,
```
Comment (mloubout) [inline devito/symbolics/extended_sympy.py:432]: ?

**Review by mloubout — COMMENTED** [review]
quick pass

File: devito/symbolics/extended_sympy.py (hunk 9 above)
Comment (FabioLuporini) [reply devito/symbolics/extended_sympy.py:432]: `np.int` is deprecated in numpy, you should use `int` instead (as per warning message). Or, much better, `np.int32` or `np.int64`, as one usually does

File: devito/ir/iet/nodes.py (hunk 5 above)
Comment (FabioLuporini) [reply devito/ir/iet/nodes.py:130]: lots of subclasses share this method which represents the basic cases. That allowed dropping a large chunk of redundant code across various subclasses

File: devito/core/arm.py (hunk 3 above)
Comment (FabioLuporini) [reply devito/core/arm.py:3]: nope. If you look into `linearize`, you'll see it consists of several sub-passes, among which what you call `linearize_index`. IOW, `linearize_index` is just one of the steps to implement linearization

File: devito/core/operator.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/core/operator.py:200]: oh, no, not necessarily.

File: devito/ir/iet/nodes.py [hunk 10]
```diff
@@ -273,51 +261,55 @@ def __repr__(self):
...
+                try:
+                    v = i.free_symbols
+                except AttributeError:
+                    continue
+                for s in v:
                     try:
-                        f = s.function
+                        # `try-except` necessary for e.g. Macro
+                        if isinstance(s.function, AbstractFunction):
+                            retval.append(s.function)
```
Comment (mloubout) [inline devito/ir/iet/nodes.py:284]: then why not directly
```
try:
    retval.append(s.function)
except AttributeError
```

File: devito/ir/iet/nodes.py [hunk 11]
```diff
@@ -273,51 +261,55 @@ def __repr__(self):
...
+                except AttributeError:
+                    pass
         if self.base is not None:
-            free.add(self.base)
+            retval.append(self.base)
         if self.retobj is not None:
-            free.update(self.retobj.free_symbols)
-        return free
+            retval.extend(self.retobj.free_symbols)
+        return tuple(filter_ordered(retval))
```
Comment (mloubout) [inline devito/ir/iet/nodes.py:312]: maybe be good at some point to have the `filter` function to be split into different output type because this does a lot of `list(tuple(np(tuple)` which isn't great.

File: devito/symbolics/extended_sympy.py (hunk 9 above)
Comment (mloubout) [reply devito/symbolics/extended_sympy.py:432]: IS it deprecated in the latest version or all versions we support? (>1.16)

Comment (FabioLuporini) [reply devito/symbolics/extended_sympy.py:432]: don't know honestly. I know that some CI logs contain a long list of warnings, some of which due to this thing

File: devito/ir/iet/nodes.py (hunk 11 above)
Comment (FabioLuporini) [reply devito/ir/iet/nodes.py:312]: yeah agree, though the overhead should be minimal

File: devito/ir/iet/nodes.py (hunk 10 above)
Comment (FabioLuporini) [reply devito/ir/iet/nodes.py:284]: because if it's not an AbstractFunction we don't want to collect the `.function`. The `.function` property is provided by all our `types.Basic` objects. One day we could purge some of this, but for now I need to pick the _actual_ Functions only here

Comment (FabioLuporini) [reply devito/ir/iet/nodes.py:284]: > One day we could purge some of this

I mean, the `types.Basic` hierarchy, by e.g. removing the .function from some types, but it's less trivial than it sounds I'm afraid

**Comment by FabioLuporini** [comment]
Approved by @mloubout on Slack, merging
